ooniapi/services/ooniprobe/src/ooniprobe/routers/reports.py

```diff
@@ -0,0 +1,276 @@
+    rid = generate_report_id(test_name, settings, cc, asn_i)
+    resp = OpenReportResponse(
+        backend_version="1.3.5", supported_formats=["yaml", "json"], report_id=rid
+    )
+    setnocacheresponse(response)
+    return resp
+
+
+class ReceiveMeasurementResponse(BaseModel):
+    """
+    Acknowledge
+    """
+
+    measurement_uid: str | None = Field(
+        examples=["20210208220710.181572_MA_ndt_7888edc7748936bf"], default=None
+    )
+
+
+@timer(name="receive_measurement")
+@router.post("/report/{report_id}", tags=["reports"])
+async def receive_measurement(
+    report_id: str,
+    request: Request,
+    response: Response,
+    cc_reader: CCReaderDep,
+    asn_reader: ASNReaderDep,
+    settings: SettingsDep,
+    s3_client: S3ClientDep,
+    content_encoding: str = Header(default=None),
+) -> ReceiveMeasurementResponse | Dict[str, Any]:
+    """
+    Submit measurement
+    """
+    setnocacheresponse(response)
+    empty_measurement = {}
+    try:
+        rid_timestamp, test_name, cc, asn, format_cid, rand = report_id.split("_")
+    except Exception:
+        log.info("Unexpected report_id %r", report_id[:200])
+        raise error("Incorrect format")
+
+    # TODO validate the timestamp?
+    good = len(cc) == 2 and test_name.isalnum() and 1 < len(test_name) < 30
+    if not good:
+        log.info("Unexpected report_id %r", report_id[:200])
+        error("Incorrect format")
+
+    try:
+        asn_i = int(asn)
+    except ValueError:
+        log.info("ASN value not parsable %r", asn)
+        error("Incorrect format")
+
+    if asn_i == 0:
+        log.info("Discarding ASN == 0")
+        Metrics.MSMNT_DISCARD_ASN0.inc()
+        return empty_measurement
+
+    if cc.upper() == "ZZ":
+        log.info("Discarding CC == ZZ")
+        Metrics.MSMNT_DISCARD_CC_ZZ.inc()
+        return empty_measurement
+
+    data = await request.body()
+    if content_encoding == "zstd":
+        try:
+            data = zstd.decompress(data)
+            ratio = len(data) / len(data)
+            log.debug(f"Zstd compression ratio {ratio}")
+        except Exception as e:
+            log.info("Failed zstd decompression")
+            error("Incorrect format")
+
+    # Write the whole body of the measurement in a directory based on a 1-hour
+    # time window
+    now = datetime.now(timezone.utc)
+    h = sha512(data).hexdigest()[:16]
+    ts = now.strftime("%Y%m%d%H%M%S.%f")
+
+    # msmt_uid is a unique id based on upload time, cc, testname and hash
+    msmt_uid = f"{ts}_{cc}_{test_name}_{h}"
+    Metrics.MSMNT_RECEIVED_CNT.inc()
+
+    compare_probe_msmt_cc_asn(cc, asn, request, cc_reader, asn_reader)
+    # Use exponential back off with jitter between retries
+    N_RETRIES = 3
+    for t in range(N_RETRIES):
+        try:
+            url = f"{settings.fastpath_url}/{msmt_uid}"
+            # TODO would be nice to make this async due to the size of the data being transmitted.
+            urlopen(url, data, 59)
```

Comment:
I think we might have to make this async otherwise we will have issues with these requests blocking the API thread.

Reply:
Agree, I used httpx as async client since it was already installed in ooniprobe